### PR TITLE
[Tool] Support validate tools

### PR DESCRIPTION
**Pull request by lalala123123**
# Description
- Tool validate CLI:
Help:
image

Validate succeed:
image

Validate failed:
image

- SDK experience:
```
from package_name import tool_func 

client = PFClient()
validation_result = client.tools.validate(source=tool_func, raise_error=False) 

if not validation_result.passed: 
    for name, error_message in validation_result.error_messages.items(): 
        print(error_message) 
```
validate result:
image

raise exception:
image

- List package tool
It will print warning log if there are invalid tools in package.
image

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/data/tool.schema.json [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+        "python",
+        "action",
+        "prompt",
+        "custom_llm",
+        "csharp"
+      ]
+    },
+    "InputDefinition": {
+      "type": "object",
+      "additionalProperties": true,
```
Comment (lalala123123) [inline src/promptflow/promptflow/_sdk/data/tool.schema.json:123]: Change additionalProperties to true to support properties that are not defined in the input
cc @melionel

File: src/promptflow/promptflow/_cli/_pf/_tool.py [hunk 2]
```diff
@@ -151,3 +183,27 @@ def list_tool(args):
...
     print(json.dumps(package_tools, indent=4))
+
+
+@exception_handler("Tool validate")
+def validate_tool(args):
+    import importlib
+
+    pf_client = PFClient()
+    try:
+        __import__(args.source)
```
Comment (melionel) [inline src/promptflow/promptflow/_cli/_pf/_tool.py:194]: what if args.source is None ?

File: src/promptflow/promptflow/_sdk/data/tool.schema.json (hunk 1 above)
Comment (melionel) [reply src/promptflow/promptflow/_sdk/data/tool.schema.json:123]: The expected behavior is not only the InputDefinition allow additionalProperties. The whole json for the Tool should allow additionalProperties.

File: src/promptflow/promptflow/_cli/_pf/_tool.py (hunk 2 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_cli/_pf/_tool.py:194]: mark --source as requried. If user provide a invalid source, cli will raise user error.
image

**Comment by melionel** [comment]
Is there a document that guide user to manually validate theire package tools before publishing ?

**Comment by lalala123123** [comment]
> Is there a document that guide user to manually validate theire package tools before publishing ?

Added an introduction to the CLI command in this PR. I will add an introduction about tool validation in create-and-use-tool-package.md in the next PR.

File: src/promptflow/promptflow/_cli/_pf/_tool.py [hunk 3]
```diff
@@ -98,11 +99,42 @@ def add_parser_list_tool(subparsers):
...
+
+    return activate_action(
+        name="validate",
+        description="Validate tool.",
+        epilog=epilog,
+        add_params=[
+            add_param_source,
+        ],
+        subparsers=subparsers,
+        help_message="Validate tool. Will raise error if it is not valid.",
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_cli/_pf/_tool.py:126]: Will raise error or return error?

File: src/promptflow/promptflow/_cli/_pf/_tool.py [hunk 4]
```diff
@@ -98,11 +99,42 @@ def add_parser_list_tool(subparsers):
     )
 
 
+def add_parser_validate_tool(subparsers):
+    """Add tool list parser to the pf tool subparsers."""
+    epilog = """
+Examples:
+
+# Validate single function tool:
+pf tool validate -–source <package_name>.<module_name>.<tool_function>
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_cli/_pf/_tool.py:108]: @D-W- align the entry design here

File: src/promptflow/promptflow/_cli/_pf/_tool.py (hunk 3 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_cli/_pf/_tool.py:126]: If an invalid tool is included, it will exit 1

**Review by melionel — DISMISSED** [review]

File: src/promptflow/promptflow/_sdk/_pf_client.py [hunk 5]
```diff
@@ -182,6 +182,11 @@ def runs(self) -> RunOperations:
         """Run operations that can manage runs."""
         return self._runs
 
+    @property
+    def tools(self) -> ToolOperations:
+        """Run operations that can manage runs."""
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_pf_client.py:187]: update docstrings

File: src/promptflow/promptflow/_sdk/operations/_tool_operations.py [hunk 6]
```diff
@@ -1,38 +1,135 @@
...
 
 class ToolOperations:
     """ToolOperations."""
 
-    def generate_tool_meta(self, tool_module):
+    def __init__(self):
+        self._tool_schema = None
+
+    @property
+    def tool_schema(self):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/operations/_tool_operations.py:44]: is this private?

File: src/promptflow/promptflow/_sdk/operations/_tool_operations.py [hunk 7]
```diff
@@ -189,3 +419,55 @@ def list(
             tools = {"package": {}, "code": {}}
         tools["package"].update(package_tools)
         return tools
+
+    @monitor_operation(activity_name="pf.tools.validate", activity_type=ActivityType.PUBLICAPI)
+    def validate(
+        self, source: Union[str, callable, PathLike], *, raise_error: bool = False, **kwargs
+    ) -> ValidationResult:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/operations/_tool_operations.py:426]: ValidationResult should be public exposed？

File: src/promptflow/promptflow/_sdk/_pf_client.py (hunk 5 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/_pf_client.py]: fix it

File: src/promptflow/promptflow/_sdk/operations/_tool_operations.py (hunk 6 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/operations/_tool_operations.py]: change to private

File: src/promptflow/promptflow/_sdk/operations/_tool_operations.py (hunk 7 above)
Comment (lalala123123) [reply src/promptflow/promptflow/_sdk/operations/_tool_operations.py:426]: Yes, ValidationResult  needs to be exposed to user

**Review by wangchao1230 — APPROVED** [review]

**Review by D-W- — APPROVED** [review]

**Pull request merged**